opm/simulators/linalg/FlexibleSolver.hpp

```diff
@@ -0,0 +1,175 @@
+/*
+  Copyright 2019 SINTEF Digital, Mathematics and Cybernetics.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef OPM_FLEXIBLE_SOLVER_HEADER_INCLUDED
+#define OPM_FLEXIBLE_SOLVER_HEADER_INCLUDED
+
+#include <opm/simulators/linalg/PreconditionerFactory.hpp>
+
+#include <dune/common/fmatrix.hh>
+#include <dune/istl/bcrsmatrix.hh>
+#include <dune/istl/solvers.hh>
+#include <dune/istl/umfpack.hh>
+
+#include <boost/property_tree/ptree.hpp>
+
+namespace Dune
+{
+
+
+template <class X, class Y>
+class SolverWithUpdate : public Dune::InverseOperator<X, Y>
+{
+public:
+    virtual void updatePreconditioner() = 0;
+};
+
+
```

Comment:
The interface here is for the call on line 134 of ISTLSolverEbosFlexible.hpp. Since that class uses `FlexibleSolver` directly, this interface is superfluous and can be removed.

It remains however a question of how we would like to expose this functionality in general. A more dune-like way would perhaps be to have the preconditioner in the same scope as the solver, and call update() on it directly, but part of what FlexibleSolver does is encapsulating the creation of the operator, scalar product and preconditioner, since they are connected (parallel vs serial), and I think that is nice. Perhaps adding a preconditioner() accessor and calling solver.preconditioner().update() would be better?